Implementing s3

joeylouise opened the pull request:
This PR does 3 things

- When user adds memory, pushes to S3
- When view memory, pulls from S3
- When deleting, deletes from S3

#116 #125

pbywater committed:
adding endpoint for creating photo memory

relates to #116

pbywater committed:
adding xhr requests

relates to #116

pbywater committed:
adding signs3 endpoint

relates to #116

pbywater committed:
woop woop photos uploading yay

relates to #116

alice-carr committed:
added getUrl func
relates #116

alice-carr committed:
added heading and tag button to photo
relates #116

alice-carr committed:
merging with master

alice-carr committed:
updating index html
relates #116

joeylouise committed:
Remove not null from heading and tag in memories table
Relates #116

joeylouise committed:
Save photo name in local storage and send ajax request to update db with heading and tag
Relates #116

joeylouise committed:
Remove comments from db_create
Relates #116

joeylouise committed:
Add updatePhotoMemory function to update tag and heading in db
Relates #116

joeylouise committed:
Call updatePhotoMemory at memory-input-photo
Relates #116

joeylouise committed:
Call createMemory at signs3 route to add image url to the db
Relates #116

joeylouise committed:
Add put route for memory-input-photo
relates #116

joeylouise committed:
Add classes to heading and tag inputs
Relates #116

joeylouise committed:
Remove addedImage from localStorage when menu close button is clicked
Relates #116

joeylouise committed:
Change test data in db build so that images are stored in memory_asset_url
Relates #116

joeylouise committed:
Add ajax request to get memory_url from s3 and append url to src of image
Relates #116

joeylouise committed:
Add new get route for memory-input-photo and refactor routes
Relates #116

joeylouise committed:
Create route for getting photo url from s3
relates #116

joeylouise committed:
Define getMemoryUrlById function that uses id from memory node to get memory_asset_url
relates #116

joeylouise committed:
Call getMemoryUrlById and delete relevant memory_asset_url from s33
relates #116

joeylouise committed:
Merge branch 'master' into implementing-s3

joeylouise committed:
Removing comments from main.js

joeylouise committed:
Remove console log from db_get.js

joeylouise committed:
Delete unnecessary helper file
relates #116 #125

joeylouise committed:
Merge branch 'implementing-s3' of github.com:pbywater/echo into implementing-s3

des-des commented on public/scripts/s3-uploads.js:
```diff
@@ -0,0 +1,61 @@
+    const files = document.getElementById('camera-input').files;
+    const file = files[0];
+    if (file == null) {
+      return alert('No file selected.');
```
? Is this good UX?

des-des commented on public/scripts/s3-uploads.js:
```diff
@@ -0,0 +1,61 @@
+function getSignedRequest(file) {
+  const xhr = new XMLHttpRequest();
+  xhr.open('GET', `/sign-s3?file-name=${file.name}&file-type=${file.type}`);
+  xhr.onreadystatechange = () => {
```
Create a function in helpers called request. Use it here

des-des commented on public/scripts/s3-uploads.js:
```diff
@@ -0,0 +1,61 @@
+        uploadFile(file, response.signedRequest, response.url);
+        saveInLocalStorage(response.key);
+      } else {
+        alert('Could not get signed URL.');
```
Not sure about these alerts ...

des-des commented on public/scripts/s3-uploads.js:
```diff
@@ -0,0 +1,61 @@
+}
+
+function uploadFile(file, signedRequest, url) {
+  const xhr = new XMLHttpRequest();
```
With a helper

des-des commented on public/scripts/s3-uploads.js:
```diff
@@ -0,0 +1,61 @@
+
+document.getElementById('photo-save').onclick = function (e) {
+  e.preventDefault();
+  if (localStorage.getItem('addedImage')) {
```
Okay, this key represents if an upload is in progress.

Two things.
  1. I do not see why you need local storage for this.
  2. Is there a case where you are uploading more than 1 image / asset. Not really sure how you want to handle this
  3. You should call this imageUploadPending, and set it to true.

joeylouise replied:
I've removed where we use localStorage and instead are using `imageUploadPending` variable

des-des commented on public/scripts/s3-uploads.js:
```diff
@@ -0,0 +1,61 @@
+      method: 'PUT',
+      url: 'memory-input-photo',
+      data: { tag, heading, imageId },
+      success: localStorage.removeItem('addedImage'),
```
I expect you mean
```js
success: () => localStorage.removeItem('addedImage')
```

joeylouise replied:
I've changed this to

`success: () => { imageUploadPending = false; }`

des-des commented on src/database/db_get.js:
```diff
@@ -43,8 +43,16 @@ const getUser = (input, callback) => {
   });
 };
 
+const getMemoryUrlById = (id, callback) => {
```
Why cant you just use getMemoryById?

joeylouise replied:
Have renamed this to getMemoryById

des-des commented on src/database/db_update.js:
```diff
@@ -5,6 +5,15 @@ const updateLikes = (likeNum, memoryId, cb) => {
+  connect.query(
+    `UPDATE memories
+     SET tag = $2, heading = $3
+     WHERE user_id = (SELECT users.id FROM users WHERE username = $1 OR email = $1)
```
```sql
WHERE user_id = (SELECT users.id FROM users WHERE username = $1 OR email = $1)
```
is the same as
```sql
WHERE username = $1 OR email = $1
```
?

I am fairly confused, this set the heading and tag of a memory, using it image id and user id?

des-des replied:
Hmm, why cant you just store the user Id in the session?

joeylouise replied:
Cool, now storing user id in session as well
#152

joeylouise replied:
Passing in user.id now so have changed query to 

`WHERE user_id = $1`

des-des commented on src/server/controllers/addPhotoMemory.js:
```diff
@@ -0,0 +1,7 @@
+const { updatePhotoMemory } = require('./../../database/db_update');
+
+module.exports = (req, res) => {
+  updatePhotoMemory('test', req.body.tag, req.body.heading, req.body.imageId, (err, res) => {
```
Your endpoints are authenticated??

This is problematic

joeylouise replied:
Have changed test to `req.session.id`
Need to add authentication to all routes

des-des commented on src/server/controllers/getPhotoMemory.js:
```diff
@@ -0,0 +1,23 @@
+const S3_BUCKET_NAME = process.env.S3_BUCKET_NAME;
+
+module.exports = (req, res) => {
+  aws.config.update({ accessKeyId: process.env.AWSAccessKeyId, secretAccessKey: process.env.AWSSecretKey, signatureVersion: 'v4', region: 'eu-west-2' });
```
Bring this out into its own helper function

joeylouise replied:
I've created an `s3init` function

des-des commented on src/server/controllers/getPhotoMemory.js:
```diff
@@ -0,0 +1,23 @@
+
+  const s3Params = {
+    Bucket: S3_BUCKET_NAME,
+    Key: req.query.memoryUrl,
```
Okay, there are 2 things you are missing here.

Firstly, you need to check if this memory belongs to the user!

I would

  1. Only keep the memoryId on the client (not the s3 key, helpfully called memoryUrl)
  2. In the endpoint, use the id from the session (this means that the user has been authenticated), and the memoryId, to find a the s3key for the memory.
  3. Make sure you check the memory belongs to the authenticated user!

joeylouise replied:
k have refactored this now so that memory_asset_url is not kept on client

des-des commented on src/server/controllers/getPhotoMemory.js:
```diff
@@ -0,0 +1,23 @@
+  s3.getSignedUrl('getObject', s3Params, (err, url) => {
+    if (err) {
+      console.log(err);
+      return res.end();
```
this is unhelpful. what does the client do??
```js
res.status(500).send('failed to generate upload url')
```

joeylouise replied:
Added `res.status(500).send('failed to generate asset URL')`

des-des commented on src/server/controllers/signS3.js:
```diff
@@ -0,0 +1,49 @@
+const S3_BUCKET_NAME = process.env.S3_BUCKET_NAME;
+
+module.exports = (req, res) => {
+  aws.config.update({ accessKeyId: process.env.AWSAccessKeyId, secretAccessKey: process.env.AWSSecretKey, signatureVersion: 'v4', region: 'eu-west-2' });
```
Again, put the s3 init logic in a helper

joeylouise replied:
I've created an `s3init` function

des-des commented on src/server/controllers/getPhotoMemory.js:
```diff
@@ -0,0 +1,23 @@
+
+  const s3Params = {
+    Bucket: S3_BUCKET_NAME,
+    Key: req.query.memoryUrl,
```
Okay,

  1. This is the s3 key, you should not give this to the client.
  2. The client should send the memoryId
  3. Use the memoryId to find the s3 key
  4. **Check that this memory belongs to the user**. Ie query the memories against the memoryId **and** the userId, to retrieve the s3Key

joeylouise replied:
have done this now

des-des commented on src/server/controllers/getPhotoMemory.js:
```diff
@@ -0,0 +1,23 @@
+  s3.getSignedUrl('getObject', s3Params, (err, url) => {
+    if (err) {
+      console.log(err);
+      return res.end();
```
res.status(500).send('failed to generate asset URL')

joeylouise replied:
I've added this in

des-des commented on src/server/controllers/signS3.js:
```diff
@@ -0,0 +1,49 @@
+const express = require('express');
```
Why is this called signS3, this is confusing.

I think you should call the files

`assetCreateUrl`, `assetGetUrl`, `assetDeleteUrl`.

Right now the naming is all over the place

des-des commented on src/server/controllers/signS3.js:
```diff
@@ -0,0 +1,49 @@
+const S3_BUCKET_NAME = process.env.S3_BUCKET_NAME;
+
+module.exports = (req, res) => {
+  aws.config.update({ accessKeyId: process.env.AWSAccessKeyId, secretAccessKey: process.env.AWSSecretKey, signatureVersion: 'v4', region: 'eu-west-2' });
```
Put this in a helper..

joeylouise replied:
I've created an `s3init` function

des-des commented on src/server/controllers/signS3.js:
```diff
@@ -0,0 +1,49 @@
+    tag: '',
+  };
+
+  createMemory('test', newMemory, 'image', fileKey, (err, res) => {
```
Hmm, this is not how you do async in js..
```js
(res, res) => {
  db.createMemory(res.body, err => {
    if (err) return res.status(500).send(message bla)
    s3.getSignedUrl(() => {
      res.success ..
    })
  })
}
```js

joeylouise replied:
I've refactored this 👍

des-des commented on src/server/controllers/signS3.js:
```diff
@@ -0,0 +1,49 @@
+  s3.getSignedUrl('putObject', s3Params, (err, data) => {
+    if (err) {
+      console.log(err);
+      return res.end();
```
Why no status code, no message ..

joeylouise replied:
Added status code and message

des-des commented on src/server/controllers/signS3.js:
```diff
@@ -0,0 +1,49 @@
+      signedRequest: data,
+      url: `https://${S3_BUCKET_NAME}.s3.amazonaws.com/${fileName}`,
+    };
+    res.write(JSON.stringify(returnData));
```
```js
res.json(returnData)
```

joeylouise replied:
I've changed this

des-des reviewed (changes requested):
okay, need some work ...

But yeah congrats getting this working, I know its a real pain

joeylouise committed:
Create s3init function which initialises s3 with aws
relates #116

joeylouise committed:
Call s3init in deleteMemory
relates #116 #125

joeylouise committed:
Call s3init in getPhotoMemory
relates #116 #125

joeylouise committed:
Call s3init in signS3
relates #116 #125

joeylouise committed:
Set imageUploadPending to false instead of removing localStorage
relates #116 #125

joeylouise committed:
Add RETURNING id to createMemory so that memory id is available
relates #116 #125

joeylouise committed:
Check id instead of memory_asset_url when updating tag and heading
relates #116 #125

joeylouise committed:
Create imageUploadPending variable
Set this to true while file is uploading
Remove saveInLocalStorage function
Create updateTagAndHeading function and pass imageId in to this
Set imageUploadPending to false once memory tag and heading have been saved in db
Relates #116 #125

joeylouise committed:
Move s3.getSignedUrl function inside createMemory so that response.rows[0].id is available
Relates #116 #125

joeylouise committed:
Rename getMemoryUrlById to getMemoryById
relates #116

joeylouise committed:
Rename getMemoryUrlById to getMemoryById
relates #116

joeylouise committed:
Return data as res.json instead of res.write(JSON.stringify...
relates #116

joeylouise committed:
Add status code and error message if fail to generate asset url
relates #116

joeylouise committed:
Change query as we are now passing in user.id so can say WHERE user_id = $1
relates #116 #152

joeylouise committed:
Pass req.session.id in to updatePhotoMemory instead of req.session.name
relates #116 #152

joeylouise committed:
Set user.id in session when user logs in
relates #152

joeylouise committed:
Remove console.log from getMemoryById
relates #116

joeylouise committed:
Add status code and error message if error
relates #116

joeylouise committed:
Pass id in to createMemory as id is now available in req.session
relates #152

joeylouise committed:
Pass req.session.id in to createMemory
relates #152

joeylouise committed:
Pass req.session.id in to createMemory
relates #152

des-des commented:
@joeylouise push?

joeylouise commented:
@des-des ok...have pushed some stuff, still working on it

des-des commented:
ah cool

joeylouise committed:
Merge branch 'master' into implementing-s3

joeylouise committed:
Remove memory_asset_url from nodes and memories array (do not want to pass this to the client)
relates #154 #116

joeylouise committed:
Pass userId in to getMemoryById so that only authenticated user is getting memories
relates #116

joeylouise committed:
Pass req.session.id in when calling getMemoryById
relates #116

joeylouise committed:
Pass req.session.name in to getMemories instead of "test"
relates #116

joeylouise committed:
Call getMemoryById in getPhotoMemory in order to query db for memory_asset_url
relates #116

joeylouise committed:
Pass memoryId into GET request at memory-input-photo rather than memory_asset_url
relates #116

joeylouise committed:
Merge branch 'implementing-s3' of github.com:pbywater/echo into implementing-s3

joeylouise committed:
Merge branch 'master' into implementing-s3

joeylouise committed:
Pass userLogin to getMemories instead of test
relates #152

joeylouise committed:
Send error or success status codes in callback in updatePhotoMemory
relates #116

joeylouise committed:
Rename files that deal with s3 urls
relates #116

joeylouise committed:
Rename files that deal with s3 urls
relates #116

joeylouise commented:
Hey @des-des I've made a bunch of changes to this, there are a couple of things I'm not sure about (making the xhr function in a helper and what to change the `alert`s to. Shall I assign to you to review so far or should we have a chat first?

des-des commented:
@joeylouise since we are both in the space lets just chat. Come grab me whenever

joeylouise committed:
Remove alerts in s3 functions and move xhr.open to bottom of xhr request functions
relates #116

joeylouise committed:
Merge branch 'master' into implementing-s3

joeylouise commented:
Hey @des-des, I tried to refactor the xhr requests and everything kept breaking and I didn't want to take any more time refactoring so I have refactored a little bit and got rid of all alerts...

des-des commented on src/database/db_get.js:
```diff
@@ -43,8 +43,16 @@ const getUser = (input, callback) => {
 
+const getMemoryById = (userId, id, callback) => {
+  connect.query('SELECT memory_asset_url FROM memories WHERE user_id = $1 AND id = $2', [userId, id], (err, res) => {
+    if (err) return err;
```
:scream: the callback is not being called when you have an error

des-des commented on src/server/app.js:
```diff
@@ -5,6 +5,9 @@ const path = require('path');
 const bodyParser = require('body-parser');
 const cookieSession = require('cookie-session');
 const router = require('./router');
+const aws = require('aws-sdk');
```
does this do anything?

des-des commented on src/server/app.js:
```diff
@@ -5,6 +5,9 @@ const path = require('path');
 const router = require('./router');
+const aws = require('aws-sdk');
+
+const S3_BUCKET_NAME = process.env.S3_BUCKET;
```
and this?

des-des commented on src/server/controllers/assetCreateUrl.js:
```diff
@@ -0,0 +1,47 @@
+        url: `https://${S3_BUCKET_NAME}.s3.amazonaws.com/${fileName}`,
+      };
+      res.json(returnData);
+      res.end();
```
Do you need to call res.json and res.end?

des-des commented on src/server/controllers/assetDeleteUrl.js:
```diff
@@ -0,0 +1,35 @@
+    });
+  });
+
+  deleteMemory(req.body.id, (error, response) => {
```
You should nest all you async stuff, right now you are running s3.delete object and deleteMemory in parallel.

Put one inside the other, or use async.parallel.

des-des commented on src/server/controllers/assetDeleteUrl.js:
```diff
@@ -0,0 +1,35 @@
+
+    s3.deleteObjects(params, (err, data) => {
+      if (err) console.log(err);
+      else console.log(data);
```
not sure you want to log this

joeylouise committed:
Remove unnecessary requires in app.js
relates #116

joeylouise committed:
Refactor getMemoryById so that callback called with error
relates #116

joeylouise committed:
Remove res.end() as res.json() is all that is needed here
relates #116

joeylouise committed:
Pass userId as parameter in deleteMemory function so that userId is checked before deleting
relates #116 #125

joeylouise committed:
Select media_type and memory_asset_url from memories in getMemoryById and return both columns
relates #116 #125

joeylouise committed:
Refactor /memories delete route so that delete memories is called inside of getMemoryById
relates #116 #125

joeylouise committed:
Add memoryKey variable when getMemoryById is called as response is now an object with memory_asset_url and media_type
relates #116

joeylouise commented:
@des-des cool. 👍
